### DISCUSSION TITLE
Go 语言系统监控的实现原理 | Go 语言设计与实现 · /golang-sysmon

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#593] [Announcements]
https://draveness.me/golang-sysmon

 Go 语言系统监控的实现原理

**Comment by JayenLee** [comment]
sysmon好像是不是当成了一个goroutine去调度了？因为我这边测试了一段代码发现，如果goroutine里面有一个无限的for循环，线程就会被占用，如果goroutine创建超过处理器个数后，发现所有的处理器都被占用了，并没有因为goroutine执行时间太长而把处理器抢占然后新创建线程M去执行其他goroutine

**Comment by draveness** [comment]
> @JayenLee
> sysmon好像是不是当成了一个goroutine去调度了？因为我这边测试了一段代码发现，如果goroutine里面有一个无限的for循环，线程就会被占用，如果goroutine创建超过处理器个数后，发现所有的处理器都被占用了，并没有因为goroutine执行时间太长而把处理器抢占然后新创建线程M去执行其他goroutine

你说的是 tight loop 的问题，1.13 里面 Go 语言的调度器还是不支持真正的抢占式调度的，你可以看调度器一节，里面介绍了 1.14 引入的真抢占式调度解决了这个问题。

FYI: https://draveness.me/golang/docs/part3-runtime/ch06-concurrency/golang-goroutine/#heading-3

**Comment by JayenLee** [comment]
> @draveness
> > @JayenLee
> > sysmon好像是不是当成了一个goroutine去调度了？因为我这边测试了一段代码发现，如果goroutine里面有一个无限的for循环，线程就会被占用，如果goroutine创建超过处理器个数后，发现所有的处理器都被占用了，并没有因为goroutine执行时间太长而把处理器抢占然后新创建线程M去执行其他goroutine
> 
> 你说的是 tight loop 的问题，1.13 里面 Go 语言的调度器还是不支持真正的抢占式调度的，你可以看调度器一节，里面介绍了 1.14 引入的真抢占式调度解决了这个问题。
> 
> FYI: https://draveness.me/golang/docs/part3-runtime/ch06-concurrency/golang-goroutine/#heading-3

哦，那这个sysmon就是独立在一个线程M里运行了吧，这个线程会不会参与调度？1.13有看到retake这个调用，是没有效果么？

**Comment by draveness** [comment]
> 哦，那这个sysmon就是独立在一个线程M里运行了吧，这个线程会不会参与调度？1.13有看到retake这个调用，是没有效果么？

这节应该说了这个 sysmon 在独立线程上运行

**Comment by dyxushuai** [comment]
想咨询一下 你的画图软件是什么

**Comment by draveness** [comment]
> 想咨询一下 你的画图软件是什么

看一下 https://draveness.me/sketch-and-sketch

**Comment by JayenLee** [comment]
什么时候能够再介绍下，垃圾回收机制呀

**Comment by draveness** [comment]
> 什么时候能够再介绍下，垃圾回收机制呀

很快了

**Comment by Alien-Leon** [comment]
> @draveness
> > 什么时候能够再介绍下，垃圾回收机制呀
> 
> 很快了

在博主的博客中学习了很多！十分期待博主对Go的内存管理分析😄

**Comment by draveness** [comment]
> > @draveness
> > > 什么时候能够再介绍下，垃圾回收机制呀
> > 
> > 
> > 很快了
> 
> 在博主的博客中学习了很多！十分期待博主对Go的内存管理分析😄

已发布 https://draveness.me/golang-memory-allocator

**Comment by tidyoux** [comment]
在 `抢占处理器` 小结有一处笔误：

> 2. 当处理器处于 `_Psyscall` 状态时，在满足以下两种情况下会调用 `runtime.handoffp` 让出处理器和线程的使用权

根据原代码，`handoffp` 只是让出了处理器，此时对应的线程或者阻塞在系统调用，或者绑定了某个 Goroutine（通过 `runtime.LockOSThread`）。

**Comment by qiluge** [comment]
想请教一个问题： sysmon运行的独立线程占不占GOMAXPROXS的数量？我的理解应该是不占的，是一个活跃线程之外的独立线程，由操作系统去调度？

**Comment by draveness** [comment]
> 想请教一个问题： sysmon运行的独立线程占不占GOMAXPROXS的数量？ 我的理解应该是不占的，是一个活跃线程之外的独立线程，由操作系统去调度？

这个 sysmon 我记得是没有 P 的，直接跑在 M 上

**Comment by discipline-yyl** [comment]
能否将sysmon中的日志在程序运行时到处，这样有助于分析死锁等问题？

**Comment by draveness** [comment]
> 能否将sysmon中的日志在程序运行时到处，这样有助于分析死锁等问题？

没听过这个，我也没查到

**Comment by MarkRepo** [comment]
抢占处理器

系统调用会在循环中调用 runtime.retake 函数抢占处于运行或者系统调用中的处理器，

这里有个笔误，应该是系统监控会在循环中吧

---

*2020-12-14 UPDATES: 已修复*

**Comment by Imfan** [comment]
大神，为什么  “当存在 Goroutine 处于 _Grunnable、_Grunning 和 _Gsyscall 状态时，意味着程序发生了死锁”

**Comment by draveness** [comment]
> 大神，为什么 “当存在 Goroutine 处于 _Grunnable、_Grunning 和 _Gsyscall 状态时，意味着程序发生了死锁”

这要结合上面的代码一起看，到了这段代码已经意味着当前程序中不包含正在运行的线程，所以如果这时存在正在执行的 Goroutine 是不是表示程序出了问题

**Comment by gladmo** [comment]
> runtime.newm1 会调用特定平台的 runtime.newsproc 通过系统调用 clone 创建一个新的线程并在新的线程中执行 runtime.mstart：

勘误: runtime.newsproc => runtime.newosproc

---

*2021-02-28 UPDATES: 已修复*

**Comment by Dup4** [comment]
> 该函数首先会检查 Go 语言运行时中正在运行的线程数量，我们通过调度器中的多个字段计算该值的结果：

「运行时中正在运行的线程数量」，感觉这句话不通。

**Comment by tutengdihuang** [comment]
```
	for i := 0; i < len(allgs); i++ {
		gp := allgs[i]
		if isSystemGoroutine(gp, false) {
			continue
		}
		s := readgstatus(gp)
		switch s &^ _Gscan {
		case _Gwaiting, _Gpreempted:
			grunning++
		case _Grunnable, _Grunning, _Gsyscall:
			print("runtime: checkdead: find g ", gp.goid, " in status ", s, "\n")
			throw("checkdead: runnable g")
		}
	}
```
这里1.17.2版本已经改了，
还有retake函数。
